planner/core/optimizer.go

```diff
@@ -386,6 +387,106 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 	return plan
 }
 
+// prunePhysicalColumns currently only work for MPP(HashJoin<-Exchange).
+// Here add projection instead of pruning columns directly for safety considerations.
+// And projection is cheap here for it saves the network cost and work in memory.
+func prunePhysicalColumns(sctx sessionctx.Context, plan PhysicalPlan) {
+	emptyColumns := make([]*expression.Column, 0)
+	if tableReader, ok := plan.(*PhysicalTableReader); ok {
+		if _, isExchangeSender := tableReader.tablePlan.(*PhysicalExchangeSender); isExchangeSender {
+			prunePhysicalColumnsInternal(sctx, tableReader.tablePlan, emptyColumns, false)
+		}
+	} else {
+		for _, child := range plan.Children() {
+			prunePhysicalColumns(sctx, child)
+		}
+	}
+}
+
+func (p *PhysicalHashJoin) extractUsedCols(parentUsedCols []*expression.Column) (leftCols []*expression.Column, rightCols []*expression.Column) {
+	for _, eqCond := range p.EqualConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(eqCond)...)
+	}
+	for _, neCond := range p.NAEqualConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(neCond)...)
+	}
+	for _, leftCond := range p.LeftConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(leftCond)...)
+	}
+	for _, rightCond := range p.RightConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(rightCond)...)
+	}
+	for _, otherCond := range p.OtherConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(otherCond)...)
+	}
+	lChild := p.children[0]
+	rChild := p.children[1]
+	for _, col := range parentUsedCols {
+		if lChild.Schema().Contains(col) {
+			leftCols = append(leftCols, col)
+		} else if rChild.Schema().Contains(col) {
+			rightCols = append(rightCols, col)
+		}
+	}
+	return leftCols, rightCols
+}
+
+func prunePhysicalColumnsInternal(sctx sessionctx.Context, plan PhysicalPlan, parentUsedCols []*expression.Column, meaningFulParentUsedCols bool) {
+	switch x := plan.(type) {
+	case *PhysicalHashJoin:
+		child0 := x.children[0]
+		child1 := x.children[1]
+		schemaColumns := x.Schema().Clone().Columns
+		leftCols, rightCols := x.extractUsedCols(schemaColumns)
+		prunePhysicalColumnsInternal(sctx, child0, leftCols, true)
+		prunePhysicalColumnsInternal(sctx, child1, rightCols, true)
+	case *PhysicalExchangeReceiver:
+		// Currently, prune columns only when parent node is HashJoin
+		if meaningFulParentUsedCols {
```

Comment:
`meaningFulParentUsedCols` sounds kinds of confusing to me. Is it possible to rewrite the logic as follows:
```
if x is PhysicalHashJoin {
    if left child is PhysicalExchangeReceiver and its child is PhysicalExchangeSender {
        // put a projection to prune columns if necessary
    }
    if right child is PhysicalExchangeReceiver and its child is PhysicalExchangeSender {
        // put a projection to prune columns if necessary
    }
}
```

Reply:
In this way, maybe we don't need `parentUsedCols`.

Reply:
Nice suggestion, Done.